judge/views/two_factor.py

```diff
@@ -0,0 +1,236 @@
+        self.request.session['2fa_passed'] = True
+        return self.next_page()
+
+    @classmethod
+    def render_qr_code(cls, username, key):
+        totp = pyotp.TOTP(key)
+        uri = totp.provisioning_uri(username, settings.SITE_NAME)
+
+        qr = qrcode.QRCode(box_size=1)
+        qr.add_data(uri)
+        qr.make(fit=True)
+
+        image = qr.make_image(fill_color='black', back_color='white')
+        buf = BytesIO()
+        image.save(buf, format='PNG')
+        return 'data:image/png;base64,' + base64.b64encode(buf.getvalue()).decode('ascii')
+
+
+class TOTPDisableView(TOTPView):
+    title = _('Disable Two Factor Authentication')
+    template_name = 'registration/totp_disable.html'
+
+    def check_skip(self):
+        if not self.profile.is_totp_enabled:
+            return True
+        return settings.DMOJ_REQUIRE_STAFF_2FA and self.request.user.is_staff and not self.profile.is_webauthn_enabled
+
+    def form_valid(self, form):
+        self.profile.is_totp_enabled = False
+        self.profile.totp_key = None
+        self.profile.save()
+        return self.next_page()
+
+
+class WebAuthnView(LoginRequiredMixin, View):
+    def dispatch(self, request, *args, **kwargs):
+        if not settings.WEBAUTHN_RP_ID:
+            raise Http404()
+        return super(WebAuthnView, self).dispatch(request, *args, **kwargs)
+
+
+class WebAuthnAttestationView(WebAuthnView):
+    def get(self, request, *args, **kwargs):
+        challenge = os.urandom(32)
+        request.session['webauthn_attest'] = webauthn_encode(challenge)
+        data = webauthn.WebAuthnMakeCredentialOptions(
+            challenge=challenge,
+            rp_id=settings.WEBAUTHN_RP_ID,
+            rp_name=settings.SITE_NAME,
+            user_id=request.profile.webauthn_id,
+            username=request.user.username,
+            display_name=request.user.username,
+            user_verification='discouraged',
+            icon_url=gravatar(request.user.email),
+            attestation='none',
+        ).registration_dict
+        data['excludeCredentials'] = [{
+            'type': 'public-key',
+            'id': {'_bytes': credential.cred_id},
+        } for credential in request.profile.webauthn_credentials.all()]
+        return JsonResponse(data, encoder=WebAuthnJSONEncoder)
+
+    def post(self, request, *args, **kwargs):
+        if not request.session.get('webauthn_attest'):
+            return HttpResponseBadRequest()
+
+        if 'credential' not in request.POST or len(request.POST['credential']) > 65536:
+            return HttpResponseBadRequest(_('Invalid WebAuthn response'))
+
+        if 'name' not in request.POST or len(request.POST['name']) > 100:
+            return HttpResponseBadRequest(_('Invalid name'))
+
+        credential = json.loads(request.POST['credential'])
+
+        response = webauthn.WebAuthnRegistrationResponse(
+            rp_id=settings.WEBAUTHN_RP_ID,
+            origin='https://' + request.get_host(),
```

Comment:
uhhh, don't we have a setting for this?

Reply:
What setting? This is supposed to be the scheme + hostname for the request. The same value is included in the `clientDataJSON` blob and signed. We check them for equality to make sure the response is not intended for a different domain.

Reply:
You're assuming everyone running has HTTPS, but as far as I could find there's nothing restricting webauthn usage to HTTPS-only.

Reply:
(`DMOJ_SCHEMA`, IIRC)